# Question 7

In [1]:
#part a


import numpy as np
import matplotlib.pyplot as plt
import math

def integralf(down, up, num):
    step=(up-down)/num
    j=down
    value=0
    for i in range(1, num+1):
        f1=(math.sin(10*math.sqrt(j)))**2
        f2=((math.sin(10*math.sqrt(j+step)))**2)
        value+=((f1+f2)/2)
        j+=step
    return (value*step)

N=4
a=integralf(0,1,1)
b=integralf(0,1,2)
final_int=0
while(np.abs(a-b)>1e-6):
    a=b
    b=integralf(0,1,N)
    N*=2
        
        
print(b,N)
        

0.4558324138011351 16384


In [2]:
#part b


import numpy as np
import matplotlib.pyplot as plt
import math


def integralf(down, up, num):
    step=(up-down)/num
    j=down
    value=0
    f2=(math.sin(10*math.sqrt(1)))**2
    for i in range(1, num+1):
        f1=f2
        f2=((math.sin(10*math.sqrt(j+step)))**2)
        value+=((f1+f2)/2)
        j+=step
    return (value*step)


def I(i):
    return integralf(0,1,2**(i))


def R(i,m):
    p=0
    if(m==1):
        p=I(i)
    else:
        p=((4**m)*R(i,m-1)-R(i-1, m-1))/(4**(m)-1)
    return p
    #this rule has been used from the textbook of Newman, it has a recursive mathematical proof which is involved
    
i=3    
print(R(2,1))       
while(i>2 and i<10):
    for j in range(1,i):
        print(R(i,j), end='')
    print('\n')
    i+=1
    
        

    

0.5492777218599949
0.421494884046580.4129760281923523

0.43935208707891290.44054256728106840.44098013139358766

0.45303902467955290.453951487186262270.45416432718475740.45421602991335025

0.456225110661417240.45643751639354150.456476977174609440.45648604639025590.4564882653701942

0.456504594095822670.4565232263247830.456524586799882060.456524773504294930.456524811360711440.45652482028525126

0.456289311314751370.456274959129346570.456271018380212670.456270023994253150.456269774972249240.456269712692300930.4562696971208187

0.45609122208241910.456078016133596940.456074890054299350.45607412092360950.456073929425006740.456073881599476260.456073869646165360.4560738666580296



In around 10 iterations it achieves the convergence value of the normal method which amounts to around 1024 slices, which is impressive considering we needed 16384 slices in the previous case.